ccf_specs.md

```diff
@@ -161,6 +161,91 @@ This specification uses CDDL notation to express CBOR data items:
+- Cadence strings is encoded as a tuple of type and value: the `String` type and its value.
+
+For compactness, CCF encoding skips type encoding when type can be inferred.
+
+As one example, the type of each element in `[String]` can be inferred to have type `String`.  Given this, CCF encodings of `[String]` can skip encoding each element's type of `String`.  Instead, CCF encodes `[String]` as a tuple like this: `[String]` type and a value representing a list of element values.  This avoids redundantly encoding each element's type.
+
+If type cannot be inferred, it must be encoded as part of a tuple representing a type and its value.  For example, the type of each element in `[AnyStruct]` cannot be inferred.  The type `AnyStruct` is the top type of all non-resource types and array elements can be of heterogenous types. In this case, `[AnyStruct]` is encoded as a tuple like this: `[AnyStruct]` type and a value being a list of tuples representing each element's type and value.
+
+### Valid CCF Encoding Requirements
+
+A CCF encoding is valid if it complies with "Valid CCF Encoding Requirements".
+
+Encoders MUST produce valid CCF encodings from valid input items.  Encoders are not required to check for invalid input items (e.g. invalid UTF-8 strings, duplicate dictionary keys, etc.)  Applications MUST NOT provide invalid items to encoders.
+
+Decoders MUST reject CCF encodings that are not valid unless otherwise specified in this section.
+
+A CCF encoding complies with "Valid CCF Encoding Requirements" if it complies with the following restrictions:
+
+- CBOR data items MUST be well-formed and valid as defined in RFC 8949.  E.g. CBOR text strings MUST contain valid UTF-8.  As an exception, RFC 8949 requirements for CBOR maps are not applicable because CCF does not use CBOR maps.
+
+- CCF encodings must comply with specifications in "CCF Specified in CDDL Notation" section of this document.
+
+- `composite-type.id` MUST be unique in `ccf-typedef-message` or `ccf-typedef-and-value-message`.
+
+- `composite-type.cadence-type-id` MUST be unique in `ccf-typedef-message` or `ccf-typedef-and-value-message`.
+
+- `field-name` MUST be unique in `composite-type`.
+
+- `type-ref.id` MUST refer to `composite-type.id`.
+
+- `composite-type-value.id` MUST be unique in the same `composite-type-value` data item.
+
+- `type-value-ref.id` MUST refer to `composite-type-value.id` in the same `composite-type-value` data item.
+
+- all parameter lists MUST be unique in `composite-type-value`.
+  - `name` MUST be unique in `composite-type-value.fields`. 
+  - `identifier` MUST be unique in `composite-type-value.initializers`.
+
+- elements MUST be unique in `restricted-type` or `restricted-type-value`.
+
+- keys MUST be unique in `dict-value`.  Decoders are not always required to check for duplicate dictionary keys.  In some cases, checking for duplicate dictionary key is not necessary or it may be delegated to the application.
```

Comment:
```suggestion
- Elements MUST be unique in `restricted-type` or `restricted-type-value`.

- Keys MUST be unique in `dict-value`.  Decoders are not always required to check for duplicate dictionary keys.  In some cases, checking for duplicate dictionary key is not necessary or it may be delegated to the application.
```

Reply:
Done in 92d80823ce8e46b262fe202bc81d3a3825a6a4a9.